payments/providers/cpu_ceepos.py

```diff
@@ -0,0 +1,361 @@
+        self.payload_add_customer(payload, order)
+        self.payload_add_return_and_notification_urls(payload)
+        self.payload_add_checksum(payload)
+
+        try:
+            r = requests.post(self.url_payment_api, json=payload, timeout=60)
+            r.raise_for_status()
+            return self.handle_initiate_payment_response(r.json())
+        except RequestException as e:
+            raise ServiceUnavailableError("Payment service is unreachable") from e
+
+    def payload_add_products(self, payload, order):
+        """Attaches product data to the payload
+
+        Order lines that contain bought products are retrieved through order"""
+        reservation = order.reservation
+        order_lines = OrderLine.objects.filter(order=order.id)
+        items = []
+        for order_line in order_lines:
+            product = order_line.product
+            int_tax = int(product.tax_percentage)
+            assert (
+                int_tax == product.tax_percentage
+            )  # make sure the tax is a whole number
+            items.append(
+                {
+                    "Code": product.sku,
+                    "Amount": order_line.quantity,
+                    "Price": price_as_sub_units(
+                        product.get_price_for_reservation(reservation)
+                    ),
+                    "Description": product.name,
+                    "Taxcode": int_tax,
+                }
+            )
+            payload["Products"] = items
+
+    def payload_add_customer(self, payload, order):
+        """Attaches customer data to the payload"""
+        reservation = order.reservation
+        payload.update(
+            {
+                "Email": reservation.billing_email_address,
+                "FirstName": reservation.billing_first_name,
+                "LastName": reservation.billing_last_name,
+            }
+        )
+
+    def payload_add_return_and_notification_urls(self, payload):
+        """Attaches return and notification URLs to the payload"""
+        payload.update(
+            {
+                "ReturnAddress": self.get_success_url(),
+                "NotificationAddress": self.get_notify_url(),
+            }
+        )
+
+    def payload_add_checksum(self, payload):
+        """
+        Attaches the checksum to the payload.
+
+        The checksum is calculated from a string that consists of the values of
+        the parameters in the payload and the source system specific secret key.
+        """
+        checksum_calc_values = []
+        for key, value in payload.items():
+            if isinstance(value, list):
+                for item in value:
+                    checksum_calc_values.extend(item.values())
+            else:
+                checksum_calc_values.append(value)
+        payload["Hash"] = self.calculate_checksum(checksum_calc_values)
+
+    def validate_ceepos_message_payload(self, request=None, response=None):
```

Comment:
I think it would make a cleaner interface if there was a separate function for request and response. You could also avoid some duplication easily.

Something like this to `cpu_ceepos/validating.py` (see the other comment about separating it to its own module [here](https://github.com/andersinno/respa/pull/27/files#r558042612)):
```python
def validate_ceepos_request(request, secret):
    data = request.POST if request.method == "POST" else request.GET
    return _validate(data, REQUEST_CHECKSUM_PARAMS, secret)

def validate_ceepos_response(response, secret):
    return _validate(response, RESPONSE_CHECKSUM_PARAMS, secret)

def _validate(data, params, secret):
    checksum_received = data["Hash"]
    checksum_calc_values = [data[x] for x in params if x in data]
    correct_checksum = _calculate_checksum(checksum_calc_values, secret)
    ...
```

Reply:
I refactored the code into separate functions for requests and responses, but I decided to keep the validation logic in the same module for now at least. Also the secret is now passed to the validation functions as you suggested. I believe this improved the code quality, thank you.